tests/conftest.py

```diff
@@ -1477,32 +1477,65 @@ async def async_setup_recorder_instance(
         ),
     ):
 
-        async def async_setup_recorder(
+        @asynccontextmanager
+        async def async_test_recorder(
             hass: HomeAssistant,
             config: ConfigType | None = None,
             *,
             wait_recorder: bool = True,
-        ) -> recorder.Recorder:
+        ) -> AsyncGenerator[recorder.Recorder]:
             """Setup and return recorder instance."""  # noqa: D401
             await _async_init_recorder_component(hass, config, recorder_db_url)
             await hass.async_block_till_done()
             instance = hass.data[recorder.DATA_INSTANCE]
             # The recorder's worker is not started until Home Assistant is running
             if hass.state is CoreState.running and wait_recorder:
                 await async_recorder_block_till_done(hass)
-            return instance
+            try:
+                yield instance
+            finally:
+                if instance.is_alive():
+                    await instance._async_shutdown(None)
+
+        yield async_test_recorder
+
+
+@pytest.fixture
+async def async_setup_recorder_instance(
+    async_test_recorder: RecorderInstanceGenerator,
+) -> AsyncGenerator[RecorderInstanceGenerator]:
+    """Yield callable to setup recorder instance."""
+
+    # pylint: disable-next=import-outside-toplevel
+
+    context_manager = None
+
+    async def async_setup_recorder(
+        hass: HomeAssistant,
+        config: ConfigType | None = None,
+        *,
+        wait_recorder: bool = True,
+    ) -> AsyncGenerator[recorder.Recorder]:
+        """Setup and return recorder instance."""  # noqa: D401
+
+        # Store the context manager to prevent it from being garbage collected
+        nonlocal context_manager
+        context_manager = async_test_recorder(hass, config, wait_recorder=wait_recorder)
+        # pylint: disable-next=unnecessary-dunder-call
+        return await context_manager.__aenter__()
```

Comment:
Would there be a point in cleaning up all created manages when this fixture is destructed?

```
  async with contextlib.AsyncExitStack() as manager:
      async def async_setup_recorder(...):
        return manager.enter_async_context(async_test_recorder(...))
      yield async_setup_recorder

Reply:
Yeah, that's nice, thanks!